src/inference/src/dev/core_impl.cpp

```diff
@@ -1127,29 +1136,62 @@ std::string ov::findPluginXML(const std::string& xmlFile) {
                     continue;
                 result[sub_property.first] = sub_property.second;
             }
+        };
+        if (item->first == ov::device::properties) {
+            // 1. device properties are found
+            auto& secondary_properties = item->second.as<ov::AnyMap>();
+            for (auto item2 = secondary_properties.begin(); item2 != secondary_properties.end();) {
+                auto parsed = parseDeviceNameIntoConfig(item2->first);
+                if (device == parsed._deviceName) {
+                    // 1.2 flatten the secondary property for target device
+                    update_device_property(item2->second.as<ov::AnyMap>());
+                    item2 = secondary_properties.erase(item2);
+                } else if (is_virtual_device(parsed._deviceName)) {
+                    // 1.2 keep the secondary property for the other virtual devices
+                    item2++;
+                    continue;
+                } else {
+                    // 1.3. remove the secondary property setting for other hardware device
+                    item2 = secondary_properties.erase(item2);
+                }
+            }
+            if (!secondary_properties.size())
+                item = result.erase(item);
+            else
+                item++;
+        } else if (item->first.find(ov::device::properties.name()) != std::string::npos) {
+            // 2. device properties DEVICE_PROPERTIES_<device_name_with_id> are found
+            auto parsed_name = item->first.substr(item->first.find(ov::device::properties.name()) +
+                                                  std::string(ov::device::properties.name()).length() + 1);
+            auto parsed = parseDeviceNameIntoConfig(parsed_name);
+            if (device == parsed._deviceName) {
+                // 2.1 flatten the secondary property for target device
+                update_device_property(item->second.as<ov::AnyMap>());
+            } else if (is_virtual_device(parsed._deviceName)) {
+                // 2.1 keep the secondary property for the other virtual devices but repack them
+                if (!result.count(ov::device::properties.name())) {
+                    result[ov::device::properties.name()] = ov::AnyMap{};
+                }
+                // 2.2 device properties with device name overrides device properties
+                auto& secondary_properties = result.at(ov::device::properties.name()).as<ov::AnyMap>();
+                secondary_properties.insert({parsed_name, item->second.as<ov::AnyMap>()});
+            }
             item = result.erase(item);
-        } else if (isVirtualDev) {
-            // 2. keep the secondary property for the other virtual devices
-            item++;
         } else {
-            // 3. remove the secondary property setting for other hardware device
-            item = result.erase(item);
+            // 3. Skip other properties
+            item++;
```

Comment:
general question: how much binary size for HETERO and OPENVINO libraires is increased after this change?
Asking because we introduce one more level of combinations of template classes.. need to understand the side effects.

Reply:
> general question: how much binary size for HETERO and OPENVINO libraires is increased after this change? Asking because we introduce one more level of combinations of template classes.. need to understand the side effects.

I will check it